### ISSUE TITLE
Incorrect output when transpiling to target ES6 and --prod

### ISSUE THREAD

**Issue by jgutix** [angular/angular#21948]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[X] Bug report  
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior
The bundle for production environment is outputting the following invalid code, when using the HttpClientModule
```
  return new ue.a(t=>{
    const n = this.xhrFactory.build();
    if (n.open(e.method, e.urlWithParams),
      e.withCredentials && (n.withCredentials = !0),
        e.headers.forEach((e,t)=>n.setRequestHeader(e, t.join(","))),
      e.headers.has("Accept") || n.setRequestHeader("Accept", "application/json, text/plain, */*"),
        !e.headers.has("Content-Type")) {
      const t = e.detectContentTypeHeader();
      null !== t && n.setRequestHeader("Content-Type", t)
    }
    if (e.responseType) {
      const t = e.responseType.toLowerCase();
      n.responseType = "json" !== t ? t : "text"
    }
    const r = e.serializeBody();
    let s = null;
    const o = ()=>{
      if (null !== s)
        return s;
      const t = 1223 === n.status ? 204 : n.status
        , r = n.statusText || "OK"
        , o = new zo(n.getAllResponseHeaders())
        , i = ("responseURL"in (n = n) && n.responseURL ? n.responseURL : /^X-Request-URL:/m.test(n.getAllResponseHeaders()) ? n.getResponseHeader("X-Request-URL") : null) || e.url;
      return s = new ei({
        headers: o,
        status: t,
        statusText: r,
        url: i
      })
    }
```

Especifically the error is in this line:
```
, i = ("responseURL"in (n = n) && n.responseURL ? n.responseURL : /^X-Request-URL:/m.test(n.getAllResponseHeaders()) ? n.getResponseHeader("X-Request-URL") : null) || e.url;
```
Because in the above lines the `n` was declared as constant: `const n = this.xhrFactory.build();`

Producing the following error in chrome:
```
ERROR TypeError: Assignment to constant variable.
    at o (main.afd9c695ba9c5bcbf50d.bundle.js:formatted:13612)
    at XMLHttpRequest.i (main.afd9c695ba9c5bcbf50d.bundle.js:formatted:13621)
    at e.invokeTask (polyfills.d14582d1423c077d31d1.bundle.js:1)
    at Object.onInvokeTask (main.afd9c695ba9c5bcbf50d.bundle.js:formatted:1694)
    at e.invokeTask (polyfills.d14582d1423c077d31d1.bundle.js:1)
    at t.runTask (polyfills.d14582d1423c077d31d1.bundle.js:1)
    at t.invokeTask [as invoke] (polyfills.d14582d1423c077d31d1.bundle.js:1)
    at _ (polyfills.d14582d1423c077d31d1.bundle.js:1)
    at XMLHttpRequest.k (polyfills.d14582d1423c077d31d1.bundle.js:1)
```

Similar in firefox:
```
ERROR 
TypeError: invalid assignment to const `n'
Stack trace:
o@http://localhost:4201/main.afd9c695ba9c5bcbf50d.bundle.js:1:298093
i@http://localhost:4201/main.afd9c695ba9c5bcbf50d.bundle.js:1:298355
eFQL/</</</c</e.prototype.invokeTask@http://localhost:4201/polyfills.d14582d1423c077d31d1.bundle.js:1:22948
onInvokeTask@http://localhost:4201/main.afd9c695ba9c5bcbf50d.bundle.js:1:27946
eFQL/</</</c</e.prototype.invokeTask@http://localhost:4201/polyfills.d14582d1423c077d31d1.bundle.js:1:22861
eFQL/</</</o</t.prototype.runTask@http://localhost:4201/polyfills.d14582d1423c077d31d1.bundle.js:1:18118
eFQL/</</</u</t.invokeTask@http://localhost:4201/polyfills.d14582d1423c077d31d1.bundle.js:1:24042
_@http://localhost:4201/polyfills.d14582d1423c077d31d1.bundle.js:1:34743
k@http://localhost:4201/polyfills.d14582d1423c077d31d1.bundle.js:1:34975
```
## Expected behavior
A valid code should be outputted when targeting ES6 for production

## Minimal reproduction of the problem with instructions

Given the downloadable example at the HTTP chapter of the Tour of heroes tutorial (https://angular.io/generated/zips/toh-pt6/toh-pt6.zip)
1. Disable `HttpClientInMemoryWebApiModule` and `HttpClientInMemoryWebApiModule` in `src/app/app.module.ts`, so the app try to make an actual XhttpRequest
2. Change the target to `"target": "es6",` in `tsconfig.json`
3. Run the app like: `ng serve --prod`

## What is the motivation / use case for changing the behavior?

The output from the compilation for ES6 is broken

## Environment

<pre><code>
Angular version: 5.0.0 and 5.2.2

Browser:
- [X] Chrome (desktop) version XX
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [X] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
For Tooling issues:
- Node version: 8.9.1  
- Platform:  Mac

Others:

</code></pre>
Worth noting this issue arises only with these three factors combined: 
1. Use of `HttpClientModule`
2. Target = es6
3. Building/serving with `prod`

Probably this is an issue that should be fixed in the angular-cli or uglify-webpack-plugin level, but I had no time to dig deeper.

**Comment by ethanpollitt** [comment]
Just ran into this issue today after migrating some code to use common/http, with references to old http in another part of the project (hence HttpModule and HttpClientModule both being imported).

Temporarily solved by reverting usage of common/http.

**Comment by zijianhuang** [comment]
Me 2. I have a complex business app, and a simple heros app both with HttpClient. 

With "ng build" and "ng build --aot", the apps are working fine.

With "ng build --prod", the complex business app starts fine with the login screeen, however, when making the first HttpClient call for auth, the browser console will give:
`ERROR TypeError: Assignment to constant variable.
    at r (main.14d23ee13c3a14c67d46.bundle.js:1)
    at XMLHttpRequest.a (main.14d23ee13c3a14c67d46.bundle.js:1)
    at e.invokeTask (polyfills.8e8e9e6cb9f95b7ba9ab.bundle.js:1)
    at Object.onInvokeTask (main.14d23ee13c3a14c67d46.bundle.js:1)
    at e.invokeTask (polyfills.8e8e9e6cb9f95b7ba9ab.bundle.js:1)
    at t.runTask (polyfills.8e8e9e6cb9f95b7ba9ab.bundle.js:1)
    at t.invokeTask [as invoke] (polyfills.8e8e9e6cb9f95b7ba9ab.bundle.js:1)
    at _ (polyfills.8e8e9e6cb9f95b7ba9ab.bundle.js:1)
    at XMLHttpRequest.k (polyfills.8e8e9e6cb9f95b7ba9ab.bundle.js:1)`

The http call for auth is actually successful. 

with "ng build --prod --no-aot", the complex business app could not start properly, with error:
`Uncaught TypeError: Assignment to constant variable.
    at ih._compileModule (vendor.defd7bfff770299239f9.bundle.js:1)
    at j.then (vendor.defd7bfff770299239f9.bundle.js:1)
    at Object.then (vendor.defd7bfff770299239f9.bundle.js:1)
    at ih._compileModuleAndComponents (vendor.defd7bfff770299239f9.bundle.js:1)
    at ih.compileModuleAsync (vendor.defd7bfff770299239f9.bundle.js:1)
    at useClass.compileModuleAsync (vendor.defd7bfff770299239f9.bundle.js:1)
    at ls.bootstrapModule (vendor.defd7bfff770299239f9.bundle.js:1)
    at Object.cDNt (main.98d929d5770520d2cd22.bundle.js:1)
    at o (inline.66056c9a0e270b3547d6.bundle.js:1)
    at Object.0 (main.98d929d5770520d2cd22.bundle.js:1)`

The simple heros app has similar problem with prod.

For now, I just use "ng build --aot" for production deployment.

You may find the simple heros demo app at https://github.com/zijianhuang/webapiclientgen, and locate DemoNGCli.

Stacks:
NG 5.2.2
NG Cli 1.6.6 or 1.6.7

**Comment by SelfishLucho** [comment]
I just ran into the same problem but with target:ES2015

**Comment by zijianhuang** [comment]
(cont.)
when target:es5, things become fine. I just tried this in 2 NG2+ apps. So apparently ng build --prod is having some problem with target es2015.

**Comment by slanden** [comment]
I'm having the same problem as zijianhuang with a login and auth call in my app targeting es2015. except when using the command "ng build -prod -sourcemaps" the error reads 
`ERROR TypeError: Assignment to constant variable.
    at u (http.js.pre-build-optimizer.js:1649)
    at XMLHttpRequest.i (http.js.pre-build-optimizer.js:1771)
    at e.invokeTask (zone.js:421)
    at Object.onInvokeTask (core.js.pre-build-optimizer.js:3979)
    at e.invokeTask (zone.js:420)
    at t.runTask (zone.js:188)
    at t.invokeTask [as invoke] (zone.js:496)
    at _ (zone.js:1517)
    at XMLHttpRequest.k (zone.js:1543)`

the line where it breaks in that file reads 
`function defaultErrorLogger(console, ...values) {
    (/** @type {?} */ (console.error))(...values);`

**Comment by zijianhuang** [comment]
At this stage, I am not entire sure if the cause is in Angular/HttpClient or in Angular CLI. So I had reported to https://github.com/angular/angular-cli/issues/9495 as well.

**Comment by MariuszHenn** [comment]
you can dirty patch (replace xhr with _xhr)
```/node_modules/@angular/common/esm2015/http.js:1649```
and replace xhr with _xhr   :P

source code reference: 
https://github.com/angular/angular/blob/cf8d512e431cdff11d65c0b963ef90c5d4e2dad4/packages/common/http/src/xhr.ts#L24

**Comment by johan--** [comment]
Same issue here, resolved by using es5 as target.

**Comment by MariuszHenn** [comment]
I had the same issue, im not sure why, but removing node_modules and sets packages as below fixed it (target es6) 
```
Angular CLI: 1.6.8
Node: 8.9.1
OS: darwin x64
Angular: 5.2.3
... common, compiler, compiler-cli, core, forms, http
... language-service, platform-browser, platform-browser-dynamic
... router, service-worker

@angular/cli: 1.6.8
@angular-devkit/build-optimizer: 0.3.1
@angular-devkit/core: 0.0.29
@angular-devkit/schematics: 0.0.52
@ngtools/json-schema: 1.1.0
@ngtools/webpack: 1.9.8
@schematics/angular: 0.1.17
typescript: 2.6.2
webpack-dev-server: 2.11.1
webpack: 3.10.0
```

**Comment by jameskentTX** [comment]
I had the same problem, when target was ES2015. But app compiles and works with:

"target": "es5" - in tsconfig.json
"typescript": "~2.5.3" - in package.json
I'm using "@angular/cli": "^1.7.0", "@angular/compiler": "^5.2.5"

**Comment by zijianhuang** [comment]
As of CLI 1.7.1 and the Angular 5.2.6, the problem still persists.

**Comment by trotyl** [comment]
@zijianhuang See https://github.com/angular/angular-cli/pull/9699, it's not an Angular issue and has been fixed in [Angular CLI v6.0.0-beta.3](https://github.com/angular/angular-cli/releases/tag/v6.0.0-beta.3).

**Comment by DcsMarcRemolt** [comment]
I use @ngtools/webpack@1.10.0 and @angular-devkit/build-optimizer@0.4.0 outside of an angular-cli project (hand written webpack config) and I can also confirm this issue. Going down to es5 fixes it.

**Comment by trotyl** [comment]
You can use `conditionals: false` in uglify-es [`compress` option](https://github.com/mishoo/UglifyJS2/tree/harmony#compress-options) to fix it. Since it's not Angular nor Angular CLI problem, making an issue (by repro without any Angular code) to uglify-es could make it solved sooner.

**Comment by zijianhuang** [comment]
@trotyl, thanks for hint. However, in Angular CLI, there seems no way to to define the uglify-es option in .angular-cli.json, or through other means. 

Can you give more hint?

Anyway, I will report this to uglifyJS2.

**Comment by citizentim** [comment]
There's been a PR to fix this but unfortunately it didn't get merged in time for 1.7.1.  One would hope that the Angular team would consider broken ES6 output as a fairly big deal and would be forthcoming with a fix very soon.

**Comment by Cito** [comment]
Same problem here with Angular-CLI 1.7.2. When I set `conditionals: false` to `const uglifyCompressOptions = { ... }` in `node_modules/@angular/cli/models/webpack-configs/production.js`, as suggested by @trotyl, the problem disappears.

**Comment by zijianhuang** [comment]
Apparently the solution is that NG CLI provides such switch for es2015. It seems this issue could be closed here, while I had reported to https://github.com/angular/angular-cli/issues/9495.

**Comment by trotyl** [comment]
Closed by https://github.com/angular/angular-cli/pull/9889

**Comment by pkozlowski-opensource** [comment]
> Closed by angular/angular-cli#9889

Thnx @trotyl !
